src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/NuGetProjectFactory.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.VisualStudio.Shell;
+using NuGet.ProjectManagement;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    /// <summary>
+    /// Composite provider chaining calls to other providers.
+    /// </summary>
+    public class NuGetProjectFactory : IProjectSystemProvider
+    {
+        private readonly IProjectSystemProvider[] _providers;
+
+        public NuGetProjectFactory(IEnumerable<IProjectSystemProvider> providers)
+        {
+            if (providers == null)
+            {
+                throw new ArgumentNullException(nameof(providers));
+            }
+
+            _providers = providers.ToArray();
```

Comment:
MVC folks had a perf talk about memory optimizations in ASP.NET. One of the "easy take-aways" was using ToList instead of ToArray (if possible) since ToArray internally creates a list first.

---
In reply to: [84584359](https://github.com/NuGet/NuGet.Client/pull/950#discussion_r84584359) [](ancestors = 84584359)